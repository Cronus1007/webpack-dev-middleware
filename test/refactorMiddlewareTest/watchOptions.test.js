import express from 'express';

import request from 'supertest';

import middleware from '../../src';
import getCompiler from '../helpers/getCompiler';

import GetLogsPlugin from '.././helpers/GetLogsPlugin';

import webpackConfig from '.././fixtures/webpack.config';

describe('should throw an error on "run" when we watching', () => {
  let compiler;
  let getLogsPlugin;
  let instance;
  let app;
  let listen;

  function listenShorthand(done) {
    return app.listen((error) => {
      if (error) {
        return done(error);
      }

      return done();
    });
  }

  function close(done) {
    if (instance.context.watching.closed) {
      if (listen) {
        listen.close(done);
      } else {
        done();
      }
    }
  }
  beforeAll((done) => {
    compiler = getCompiler(webpackConfig);

    getLogsPlugin = new GetLogsPlugin();
    getLogsPlugin.apply(compiler);

    instance = middleware(compiler);

    app = express();
    app.use(instance);

    listen = listenShorthand(done);
  });

  afterAll(close);

  it('should logging an error', (done) => {
    request(app)
      .get('/bundle.js')
      .expect(200, (error) => {
        if (error) {
          return done(error);
        }

        return compiler.run((runError) => {
          expect(() => {
            throw runError;
          }).toThrowErrorMatchingSnapshot();

          done();
        });
      });
  });
});

describe('should throw an error on "watch" when we watching', () => {
  let compiler;
  let getLogsPlugin;
  let instance;
  let app;
  let listen;

  function listenShorthand(done) {
    return app.listen((error) => {
      if (error) {
        return done(error);
      }

      return done();
    });
  }

  function close(done) {
    if (instance.context.watching.closed) {
      if (listen) {
        listen.close(done);
      } else {
        done();
      }
    }
  }
  beforeAll((done) => {
    compiler = getCompiler(webpackConfig);

    getLogsPlugin = new GetLogsPlugin();
    getLogsPlugin.apply(compiler);

    instance = middleware(compiler);

    app = express();
    app.use(instance);

    listen = listenShorthand(done);
  });

  afterAll(close);

  it('should logging an error', (done) => {
    request(app)
      .get('/bundle.js')
      .expect(200, (error) => {
        if (error) {
          return done(error);
        }

        return compiler.watch({}, (watchError) => {
          expect(() => {
            throw watchError;
          }).toThrowErrorMatchingSnapshot();

          done();
        });
      });
  });
});

describe('should handle an earlier request if a change happened while compiling', () => {
  let instance;
  let app;
  let listen;

  function listenShorthand(done) {
    return app.listen((error) => {
      if (error) {
        return done(error);
      }

      return done();
    });
  }

  function close(done) {
    if (instance.context.watching.closed) {
      if (listen) {
        listen.close(done);
      } else {
        done();
      }
    }
  }
  beforeAll((done) => {
    const compiler = getCompiler(webpackConfig);

    instance = middleware(compiler);

    let invalidated = false;

    compiler.hooks.done.tap('Invalidated', () => {
      if (!invalidated) {
        instance.invalidate();

        invalidated = true;
      }
    });

    app = express();
    app.use(instance);

    listen = listenShorthand(done);
  });

  afterAll(close);

  it('should return the "200" code for the "GET" request to the bundle file', (done) => {
    request(app).get('/bundle.js').expect(200, done);
  });
});
