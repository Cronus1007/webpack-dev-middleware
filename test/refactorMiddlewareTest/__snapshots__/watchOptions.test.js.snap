// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw an error on "run" when we watching should logging an error 1`] = `"You ran Webpack twice. Each instance only supports a single concurrent compilation at a time."`;

exports[`should throw an error on "watch" when we watching should logging an error 1`] = `"You ran Webpack twice. Each instance only supports a single concurrent compilation at a time."`;
